text/0000-libobs-keychain-api.md

```diff
@@ -0,0 +1,69 @@
+# Summary
+
+This RFC proposes the addition of OS keychain APIs to the libobs platform utilities to faciliate secure storage of credentials.
+
+# Motivation
+
+While the credentials stored by OBS in profiles are not of severe sensitivity due to limited scope, as a modern desktop application we should still do our best to store them securely.
+
+# Detailed Changes
+
+This RFC proposes the addition of the following libobs API methods:
+
+- `bool os_keychain_save(const char *key, const char *data)`
+- `bool os_keychain_load(const char *key, char **data)`
+- `bool os_keychain_delete(const char *key)`
+
+As the API requires OS-specific implementation it shall become part of the `util/platform.h` header.
+
+Operating-system backends:
+- Windows: `wincred.h` / `Cred*` Win32 API functions
+- macOS: Keychain Services API
+- Linux: `libsecret`
+
+On some platforms we may add a prefix such as `obs-studio:` to identify keychain an entries as an OBS-related to the user. This is already common practice with many applications such as git, RDP, or Xbox Live.
+
+**Usage examples:**
+Saving:
+```c
+const char *key = "ProfileName::StreamKey";
+const char *value = "mystreamkey";
+bool success = os_keychain_save(key, value);
+```
+
+Loading:
+```c
+const char *key = "ProfileName::StreamKey";
+char *value = NULL;
+bool success = os_keychain_load(key, &value);
+
+... do stuff ...
+
+if (value)
+    bfree(value);
+```
+
+Deletion:
+```c
+const char *key = "ProfileName::StreamKey";
+bool success = os_keychain_delete(key);
+```
+
+# Drawbacks
+
+Accessing keychain information can reduce portability of profiles by not storing OAuth information or stream keys in plaintext files.
+
+In some cases accessing the keychain may require user confirmation, the API may block in such cases, which has to be accounted for by consumers.
+
+On Linux the situation may be more complicated. While both GNOME and KDE provide a "Secret Service", other desktop environments might not. This will require further investigation.  
```

Comment:
- Not KDE but Plasma or KDE Plasma, the KDE name no longer relate directly to the desktop environment
- GNOME and Plasma indeed can provide their own implementation of `org.freedesktop.secrets` (gnome-keyring and kwallet) but those are not bound to the desktop environment (e.g. kwallet can be used on GNOME).
- KeepassXC is also a provider of `org.freedesktop.secrets`

Reply:
So that would require detecting if the system has any provider for `org.freedesktop.secrets` available, and otherwise either disable any functionality requiring more sensitive credentials (for the purpose of this, I would consider stream keys to not be super sensitive), or fall back to an OBS implementation (e.g. plaintext file).